File: mf-backend/src/modules/transactions/service/transaction.service.ts
import { Injectable, NotFoundException, ForbiddenException, BadRequestException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Transaction } from '../entity/transaction.entity.js';
import { CreateTransactionDto } from '../dto/transaction.dto.js';
import { Ledger, LedgerStatus } from '../../ledgers/entity/ledger.entity.js';

@Injectable()
export class TransactionService {
  constructor(
    @InjectRepository(Transaction)
    private readonly repository: Repository<Transaction>,
    @InjectRepository(Ledger)
    private readonly ledgerRepository: Repository<Ledger>,
  ) {}

  async create(dto: CreateTransactionDto, userId: string, userName: string) {
    // Verificar se o ledger está aceito antes de permitir transações
    const ledger = await this.ledgerRepository.findOne({ where: { id: dto.ledgerId } });
    if (!ledger) throw new NotFoundException('Dívida não encontrada.');
    if (ledger.status !== LedgerStatus.ACCEPTED) {
      throw new BadRequestException('Não é possível adicionar transações a uma dívida que não foi aceita.');
    }

    const transaction = this.repository.create({
      ...dto,
      ledger: { id: dto.ledgerId },
      transactionDate: dto.transactionDate
        ? new Date(dto.transactionDate)
        : new Date(),
      createdById: userId,
      createdByName: userName,
    });
    return await this.repository.save(transaction);
  }

  async findAllByLedger(ledgerId: string) {
    return await this.repository.find({
      where: { ledger: { id: ledgerId } },
      order: { createdAt: 'DESC' },
    });
  }

  async delete(transactionId: string, userId: string) {
    const transaction = await this.repository.findOne({
      where: { id: transactionId },
      relations: ['ledger'],
    });

    if (!transaction) {
      throw new NotFoundException('Transação não encontrada.');
    }

    // Verificar se o usuário é dono ou participante do ledger
    if (
      transaction.ledger.ownerId !== userId &&
      transaction.ledger.participantId !== userId
    ) {
      throw new ForbiddenException(
        'Você não tem permissão para excluir esta transação.',
      );
    }

    await this.repository.remove(transaction);
    return { message: 'Transação excluída com sucesso.' };
  }
}

File: mf-backend/src/main/java/com/myfinance/transaction/TransactionService.java
package com.myfinance.transaction;

import com.myfinance.ledger.Ledger;
import com.myfinance.ledger.LedgerStatus;
import com.myfinance.transaction.dto.CreateTransactionDTO;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.transaction.Transactional;
import jakarta.ws.rs.BadRequestException;
import jakarta.ws.rs.ForbiddenException;

@ApplicationScoped
public class TransactionService {

    @Transactional
    public Transaction create(CreateTransactionDTO dto, String userId, String userName) {
        Ledger ledger = Ledger.findById(dto.ledgerId);

        if (ledger == null) {
            throw new BadRequestException("Conta inexistente.");
        }

        // Regra de aceite
        if (ledger.status != LedgerStatus.ACCEPTED) {
            throw new BadRequestException(
                    "Não é possível adicionar transações. A dívida está pendente de aceite ou foi recusada.");
        }

        // Validação de segurança: o Ledger pertence a quem está logado?
        if (!ledger.ownerId.equals(userId) && !userId.equals(ledger.participantId)) {
            throw new ForbiddenException("Você não tem permissão para alterar esta conta.");
        }

        Transaction transaction = new Transaction();
        transaction.ledger = ledger;
        transaction.amount = dto.amount;
        transaction.type = dto.type;
        transaction.description = dto.description;
        transaction.transactionDate = dto.transactionDate != null ? dto.transactionDate.atStartOfDay()
                : java.time.LocalDateTime.now();
        transaction.createdById = userId;
        transaction.createdByName = userName;

        transaction.persist();
        return transaction;
    }

    @Transactional
    public boolean delete(String id, String userId) {
        Transaction transaction = Transaction.findById(id);
        if (transaction == null)
            return false;

        // Recupera o ledger
        Ledger ledger = transaction.ledger;
        if (!ledger.ownerId.equals(userId) && !userId.equals(ledger.participantId)) {
            throw new ForbiddenException("Você não tem permissão para remover esta transação.");
        }

        return Transaction.deleteById(id);
    }
}
